doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,330 @@
+    size_t packet_length;
+
+    /* Pkts - might be better as one field as per ACK Manager */
+    unsigned int initial : 1;
+    unsigned int handshake : 1;
+    unsigned int rtt0 : 1;
+    unsigned int rtt1 : 1;
+    /* Pkts options */
+    PACKET_VALIDITY validity;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` struture from the ACK Manager overlaps this
+  significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in
+  the Application Data number space?
+- the `packet_header` will become a structure / union most likely.
+
+Interactions
+------------
+
+The packetiser needs to interact with other modules.  This defines the APIs
+by which it does so.
+
+### Stream Send Buffers
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+__owur int ossl_quic_packetise(SSL_QUIC_TX_PACKETISER *tx, SSL *stream,
+                               const OSSL_QUIC_FRAME *frame,
+                               size_t frame_length);
+```
+
+#### Notes
+
+- Should more than one frame be queueable at once?
+- More arguments are likely going to be required.
+- How to destroy frames?  `OPENSSL_free` or a callback?
+
+### TLS Handshake Record Layer
+
+Uses the Record Layer API to implement the inner TLS-1.3 protocol handshake.
+It produces the QUIC crypto frames which are queued using the same mechanism
+as the [Stream Send Buffers](#stream-send-buffers) above.
+
+#### Notes
+
+- How distinguish handshake from rest?
+
+### Flow Controller and Statistics Collector
+
+To make decisions about what frames to coalesce, the packetiser relies
+on the flow controller to enforce stream and connection bandwidth limits
+[RFC 9000 4.1 Data Flow Control].
+
+```c
+/* API to be defined for flow controller */
+```
+
+### Congestion Controller
+
+Also part of the frame coalescing decision is the congestion controller
+[RFC 9002].  For MVP, this will be a _just send it_.
+
+```c
+/* Pluggable congestion controller APIs go here */
+/*
+ * Pluggable congestion controller APIs go here
+ * Extract that is required from #18018
+ */
+```
+
+### QUIC Write Record Layer
+
+Coalesced frames are passed to the QUIC record layer for encryption and sending.
+To send accumulated frames as packets to the QUIC Write Record Layer:
+
+```c
+__owur int ossl_quic_send_packets(SSL_QUIC_TX_PACKETISER *tx);
+```
+
+#### Notes
+
+- Should the send call accept a stream SSL, a connection SSL or either?
+- How to handle multiple security levels in a packet?
```

Comment:
is this possible? What does the RFC say about this?

Reply:
Okay, my misinterpretation.